File: 5-Problem-A-Day/24. Swap Nodes in Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy_head = new ListNode(0);        
        dummy_head.next = head;    
        ListNode prev = null;
        ListNode curr = head;        
        while (curr != null && head.next != null) {
            if (curr.next == null) {
                prev.next = curr;
                break;
            }
            ListNode tmp = curr.next.next;            
            curr.next.next = curr;            
            if (curr == head)
                head = curr.next;
            else
                prev.next = curr.next;
            curr.next = null;
            prev = curr;
            curr = tmp;
        }
        return head;
    }
}

File: 2021/24. Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {        
        ListNode* curr = head, *prev = NULL;        
        while (curr != NULL && curr->next != NULL) {
            ListNode *neighbor = curr->next, *next = neighbor->next;
            neighbor->next = curr;
            curr->next = next;
            if (prev == NULL) {
                head = neighbor;                
            } else {
                prev->next = neighbor;
            }
            prev = curr;
            curr = next;
        }
        return head;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* dummy = new ListNode(), *prev = dummy;
        dummy->next = head;
        while (head != NULL && head->next != NULL) {
            ListNode* second = head->next, *next = second->next;            
            prev->next = second;
            second->next = head;
            head->next = next;            
            prev = head;
            head = next;
        }
        return dummy->next;
    }
};